sys/*/Kconfig: Add simple sys modules to Kconfig

MrKevinWeiss opened the pull request:
### Contribution description

Add sys based konfig modules that don't require many dependencies and can just be added.  Most are picked up by the modules wildcard from the makefile.

#### Not include but should be easy
- `MODULE_PROGRESS_BAR`

### Testing procedure

The CI should pass.

### Issues/PRs references

MrKevinWeiss commented:
@leandrolanzieri Let me know if this is useful and I will keep going.

leandrolanzieri commented on sys/analog_util/Kconfig:
```diff
@@ -0,0 +1,10 @@
+
+config MODULE_ANALOG_UTIL
+    bool "Analog utility function interfaces"
+    depends on MODULE_PERIPH_ADC
```
```suggestion
    depends on HAS_PERIPH_ADC
    select MODULE_PERIPH_ADC
```

MrKevinWeiss replied:
The thing is it only uses the headers from ADC and nothing to do with ADC stuff.  Should I just keep the same?

leandrolanzieri replied:
Ok then I think it actually does not have any dependency at all right?

leandrolanzieri commented on sys/analog_util/Kconfig:
```diff
@@ -0,0 +1,10 @@
+
+config MODULE_ANALOG_UTIL
+    bool "Analog utility function interfaces"
+    depends on MODULE_PERIPH_ADC
```
```suggestion
    depends on MODULE_PERIPH_ADC
    depends on TEST_KCONFIG
```

leandrolanzieri commented on sys/app_metadata/Kconfig:
```diff
@@ -0,0 +1,9 @@
+#
+
+config MODULE_APP_METADATA
+    bool "Application metadata printer"
```
```suggestion
    bool "Application metadata printer"
    depends on TEST_KCONFIG
```

leandrolanzieri commented on sys/base64/Kconfig:
```diff
@@ -0,0 +1,9 @@
+#
+
+config MODULE_BASE64
+    bool "Base64 Codec"
```
```suggestion
    bool "Base64 Codec"
    depends on TEST_KCONFIG
```

leandrolanzieri commented on sys/base64/Kconfig:
```diff
@@ -0,0 +1,9 @@
+# directory for more details.
+#
+
+config MODULE_BASE64
```
Looks like there is one more module here: `MODULE_BASE64URL`.

leandrolanzieri commented:
> @leandrolanzieri Let me know if this is useful and I will keep going.

Of course! :+1:

MrKevinWeiss commented:
pushed changes, should I just do a big batch in this PR or have a bit smaller chunks (I think bigger is better)

leandrolanzieri commented:
> pushed changes, should I just do a big batch in this PR or have a bit smaller chunks (I think bigger is better)

Go ahead, commits should be pretty much self-contained without so many dependencies going on.

MrKevinWeiss commented:
Have a quick look please, this is really fulfilling work :laughing: 

I will try to get some more done tomorrow but I am almost through the simple stuff.

leandrolanzieri commented on sys/base64/Kconfig:
```diff
@@ -0,0 +1,15 @@
+# directory for more details.
+#
+
+config MODULE_BASE64
+    bool "Base64 codec"
+    depends on TEST_KCONFIG
+
+config MODULE_BASE64URL
+    bool "Base64 URL codec"
+    depends on TEST_KCONFIG
+    select MODULE_BASE64
```
```suggestion
menuconfig MODULE_BASE64
    bool "Base64 codec"
    depends on TEST_KCONFIG

config MODULE_BASE64URL
    bool "Base64 URL codec"
    depends on MODULE_BASE64
```

MrKevinWeiss replied:
I think it can just select it as `MODULE_BASE64` has not dependencies.

leandrolanzieri commented on sys/cb_mux/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# directory for more details.
+#
+
+config CB_MUX
```
```suggestion
config MODULE_CB_MUX
```

MrKevinWeiss replied:
doh

leandrolanzieri commented on sys/eepreg/Kconfig:
```diff
@@ -0,0 +1,11 @@
+
+config MODULE_EEPREG
+    bool "EEPROM metadata registry"
+    depends on MODULE_PERIPH_EEPROM
```
```suggestion
    depends on HAS_PERIPH_EEPROM
    select MODULE_PERIPH_EEPROM
```

MrKevinWeiss replied:
OK will do

MrKevinWeiss commented:
Another update, this should be it (plus any fixups, maybe pay close attention to the trace module)

MrKevinWeiss commented:
I guess I can also start adding the app.test.config for each test that would work:

-pbkdf2 (maybe)
- bloom_bytes
- cb_mux_bench
- eepreg
- struct_tm_utility
- trace

All other module tests are either inside unittests or have other modules that may not be supported yet.

leandrolanzieri commented on sys/base64/Kconfig:
```diff
@@ -0,0 +1,14 @@
+
+config MODULE_BASE64URL
+    bool "Base64 URL codec"
+    select MODULE_BASE64
```
If you choose to `select`, then `MODULE_BASE64URL` `depends on TEST_KCONFIG` as well

leandrolanzieri commented on sys/oneway-malloc/Kconfig:
```diff
@@ -0,0 +1,13 @@
+    depends on TEST_KCONFIG
+    help
+        A malloc implementation without free for boards where the
+        toolchain does not implement dynamic memory allocation.
```
Missing newline

MrKevinWeiss replied:
?

leandrolanzieri replied:
Sorry, missed the fixup. I commented only looking at 4839f774a5e396a5e9bede43ffeaee31d81bf51b

leandrolanzieri commented on sys/seq/Kconfig:
```diff
@@ -0,0 +1,10 @@
+#
+
+config MODULE_SEQ
+    bool "Serial number arithmetic"
```
Better `bool "Serial number arithmetic (RFC 1982)"`

leandrolanzieri commented on sys/trace/Kconfig:
```diff
@@ -0,0 +1,19 @@
+    depends on TEST_KCONFIG
+    select MODULE_XTIMER
+
+if MODULE_TRACE
+
+config TRACE_BUFSIZE
+    int "Trace buffer size"
+    default 512
+
+endif # MODULE_TRACE
```
For this we can take the approach of #16037 and splitting the configuration option, so one can use it both while testing dependency modelling and in normal Kconfig usage.

leandrolanzieri commented on tests/trace/Makefile:
```diff
@@ -2,7 +2,4 @@ include ../Makefile.tests_common
-# reduce tracebuffer (default is 512), so this test compiles for more boards
-CFLAGS += -DCONFIG_TRACE_BUFSIZE=64
-
 include $(RIOTBASE)/Makefile.include
```
```suggestion
include $(RIOTBASE)/Makefile.include

# reduce tracebuffer (default is 512), so this test compiles for more boards
ifndef CONFIG_TRACE_BUFSIZE
  CFLAGS += -DCONFIG_TRACE_BUFSIZE=64
endif
```

leandrolanzieri commented on tests/trace/app.config.test:
```diff
@@ -0,0 +1 @@
+CONFIG_MODULE_TRACE=y
```
```suggestion
CONFIG_TRACE-BUFSIZE=64
```

leandrolanzieri commented on sys/trace/Kconfig.config:
```diff
@@ -0,0 +1,15 @@
+# directory for more details.
+#
+
+# XXX This file only is required since there is no easy way to use these config
+# options with both the final MODULE_SHELL and KCONFIG_USEMODULE_SHELL in a
+# nicely looking and easy to migrate way. After migration, the content of this
+# file can be folded back into `sys/shell/Kconfig`
```
Copy-pasta. Should say trace, not shell :-)

MrKevinWeiss replied:
Nice one, I made it more copy-pasta-able.

leandrolanzieri reviewed (commented):
Feel free to squash and trigger the CI

MrKevinWeiss commented:
It seems there is a problem with trace test not matching but I cannot reproduce it locally..

<details><summary>with kconfig</summary>

```
make clean -C tests/trace/ && TEST_KCONFIG=1 make all info-modules -C tests/trace/ && echo tests/trace/bin/native/tests_trace.bin | sha1sum | xxd -r -p | base64
make: Entering directory '/home/weiss/wd/RIOT/tests/trace'
make: Leaving directory '/home/weiss/wd/RIOT/tests/trace'
make: Entering directory '/home/weiss/wd/RIOT/tests/trace'
=== [ATTENTION] Testing Kconfig dependency modelling ===
=== [ATTENTION] Testing Kconfig dependency modelling ===
Building application "tests_trace" for "native" with MCU "native".

"make" -C /home/weiss/wd/RIOT/boards/native
"make" -C /home/weiss/wd/RIOT/boards/native/drivers
"make" -C /home/weiss/wd/RIOT/core
"make" -C /home/weiss/wd/RIOT/cpu/native
"make" -C /home/weiss/wd/RIOT/cpu/native/periph
"make" -C /home/weiss/wd/RIOT/cpu/native/stdio_native
"make" -C /home/weiss/wd/RIOT/drivers
"make" -C /home/weiss/wd/RIOT/drivers/periph_common
"make" -C /home/weiss/wd/RIOT/sys
"make" -C /home/weiss/wd/RIOT/sys/auto_init
"make" -C /home/weiss/wd/RIOT/sys/div
"make" -C /home/weiss/wd/RIOT/sys/test_utils/interactive_sync
"make" -C /home/weiss/wd/RIOT/sys/trace
"make" -C /home/weiss/wd/RIOT/sys/xtimer
   text	   data	    bss	    dec	    hex	filename
  32878	    624	  48380	  81882	  13fda	/home/weiss/wd/RIOT/tests/trace/bin/native/tests_trace.elf
auto_init
auto_init_xtimer
board
core
core_idle_thread
core_init
core_msg
core_panic
cpu
div
native_drivers
periph
periph_common
periph_gpio
periph_gpio_linux
periph_init
periph_init_gpio
periph_init_gpio_linux
periph_init_pm
periph_init_timer
periph_init_uart
periph_pm
periph_timer
periph_uart
stdin
stdio_native
sys
test_utils_interactive_sync
trace
xtimer
make: Leaving directory '/home/weiss/wd/RIOT/tests/trace'
8vwq6UgXquh6q880BCFyV5Q7W78=
```
</details>

<details><summary>without kconfig</summary>

```
make clean -C tests/trace/ && TEST_KCONFIG=0 make all info-modules -C tests/trace/ && echo tests/trace/bin/native/tests_trace.bin | sha1sum | xxd -r -p | base64
make: Entering directory '/home/weiss/wd/RIOT/tests/trace'
make: Leaving directory '/home/weiss/wd/RIOT/tests/trace'
make: Entering directory '/home/weiss/wd/RIOT/tests/trace'
Building application "tests_trace" for "native" with MCU "native".

"make" -C /home/weiss/wd/RIOT/boards/native
"make" -C /home/weiss/wd/RIOT/boards/native/drivers
"make" -C /home/weiss/wd/RIOT/core
"make" -C /home/weiss/wd/RIOT/cpu/native
"make" -C /home/weiss/wd/RIOT/cpu/native/periph
"make" -C /home/weiss/wd/RIOT/cpu/native/stdio_native
"make" -C /home/weiss/wd/RIOT/drivers
"make" -C /home/weiss/wd/RIOT/drivers/periph_common
"make" -C /home/weiss/wd/RIOT/sys
"make" -C /home/weiss/wd/RIOT/sys/auto_init
"make" -C /home/weiss/wd/RIOT/sys/div
"make" -C /home/weiss/wd/RIOT/sys/test_utils/interactive_sync
"make" -C /home/weiss/wd/RIOT/sys/trace
"make" -C /home/weiss/wd/RIOT/sys/xtimer
   text	   data	    bss	    dec	    hex	filename
  32878	    624	  48380	  81882	  13fda	/home/weiss/wd/RIOT/tests/trace/bin/native/tests_trace.elf
auto_init
auto_init_xtimer
board
core
core_idle_thread
core_init
core_msg
core_panic
cpu
div
native_drivers
periph
periph_common
periph_gpio
periph_gpio_linux
periph_init
periph_init_gpio
periph_init_gpio_linux
periph_init_pm
periph_init_timer
periph_init_uart
periph_pm
periph_timer
periph_uart
stdin
stdio_native
sys
test_utils_interactive_sync
trace
xtimer
make: Leaving directory '/home/weiss/wd/RIOT/tests/trace'
8vwq6UgXquh6q880BCFyV5Q7W78=
```
</details>

MrKevinWeiss commented:
rebased

MrKevinWeiss commented:
Dropped trace due to complications with murdock and kconfig race conditions due to the config parameter.

This will get corrected and added in a different PR.

fjmolinas commented:
> Dropped trace due to complications with murdock and kconfig race conditions due to the config parameter.
> 
> This will get corrected and added in a different PR.

Can you elaborate? what race conditions? Would these apply elsewhere as well?

MrKevinWeiss commented:
Ahh sorry, @leandrolanzieri is working on it now.

[murdock](https://ci.riot-os.org/RIOT-OS/RIOT/16009/40e6e92225903f313a12313aca1012148f87ac8c/output/compile/tests/trace/native:gnu.txt) failed the `tests/trace` because the `CONFIG_TRACE_BUFSIZE` remained at 512 and not at 64 when using `TEST_KCONFIG=0`.

running `/bin/bash -c "source .murdock; JOBS=4 compile tests/trace native:gnu"` exposes it but the commands that I used in my [comment](https://github.com/RIOT-OS/RIOT/pull/16009#issuecomment-783331842) did not see the issue.


@leandrolanzieri  understands it better, I am just poorly regurgitating the details. It turns out that you need to `make clean` before `make all`.

MrKevinWeiss commented:
Since this is already a big PR aimed at doing simple modules, it will be dealt with later (shortly) with the fix.

MrKevinWeiss commented:
Anything blocking this now?
@fjmolinas @leandrolanzieri

fjmolinas commented:
> Anything blocking this now?
> @fjmolinas @leandrolanzieri

all looks good on my side.

leandrolanzieri reviewed (approved):
Looks good! ACK

MrKevinWeiss commented:
Thanks for the review!
